// codehooks-js/index.js
import {agg} from './aggregation/index.mjs';
import {crudlify as crud} from './crudlify/index.mjs';
import {serveStatic as ws, render as renderView, internalFetch} from './webserver.mjs';
import Workflow from './workflow/engine.mjs';
import { generateOpenApiSpec } from './openapi/generator.mjs';
import { generateSwaggerHtml } from './openapi/swagger-ui.mjs';

function createRoute(str) {
  if(str instanceof RegExp) {
    return JSON.stringify({$regex: str.toString()})
  }
  return str;
}

class Codehooks {
  settings = {};
  routes = {};
  queues = {};
  jobs = {};
  auths = {};
  appmiddleware = [];
  datastore = null;
  listeners = [];
  workers = {};
  realtime = {};
  startup = {};
  openApiMeta = {};  // Store route documentation metadata

  // Extract OpenAPI metadata from route handlers
  _extractOpenApiMeta = (routeKey, hooks) => {
    for (const hook of hooks) {
      if (hook && hook._openApiSpec) {
        this.openApiMeta[routeKey] = {
          ...this.openApiMeta[routeKey],
          ...hook._openApiSpec
        };
      }
    }
  };

  post = (path, ...hook) => {
    const routeKey = `POST ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  get = (path, ...hook) => {
    const routeKey = `GET ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  put = (path, ...hook) => {
    const routeKey = `PUT ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  patch = (path, ...hook) => {
    const routeKey = `PATCH ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  delete = (path, ...hook) => {
    const routeKey = `DELETE ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  all = (path, ...hook) => {
    const routeKey = `* ${createRoute(path)}`;
    this.routes[routeKey] = hook;
    this._extractOpenApiMeta(routeKey, hook);
  };

  auth = (path, ...hook) => {
    this.auths[createRoute(path)] = hook;
  };

  use = (...hook) => {
    if ((typeof hook[0] === 'string')) {      
        this.appmiddleware.push({ path: hook[0], func: hook[1]});      
    } else if ((hook[0] instanceof RegExp)) {
      this.appmiddleware.push({ regex: JSON.stringify({$regex: hook[0].toString()}), func: hook[1]});
    }
    else {
      this.appmiddleware.push(hook);
    }
    this.appmiddleware = this.appmiddleware.flat();
  };

  useRoute = (route, ...hook) => {
    if ((typeof route === 'string')) {      
        this.appmiddleware.push({ path: route, func: hook[0]});      
    } else if ((route instanceof RegExp)) {
      this.appmiddleware.push({ regex: JSON.stringify({$regex: route.toString()}), func: hook[0]});
    }
    this.appmiddleware = this.appmiddleware.flat();
  };

  queue = (topic, ...hook) => {
    this.queues[topic] = hook;
  };

  worker = async (name, hook, options={}) => {
    this.workers[name] = [hook];
    this.queues[name] = [hook]; // legacy support
    
    // add option to key-value store if options is not empty
    if (Object.keys(options).length > 0) {
      this.set(`queue_worker:${name}`, options);
    }
  };

  job = (cronExpression, ...hook) => {
    this.jobs[cronExpression] = hook;
  }

  static = (options, hook) => {
    return ws(options, Codehooks.getInstance(), _filestore, hook, {source: true})
  }

  storage = (options, hook) => {
    return ws(options, Codehooks.getInstance(), _filestore, hook)
  }

  set = (key, val) => {
    this.settings[key] = val;
  }

  render = async (view, data, cb) => {
    renderView(view, data, this.settings, cb)
  }

  crudlify = (schema={}, options={}) => {
    return crud(Codehooks.getInstance(), schema, options);
  }

  /**
   * Enable OpenAPI documentation
   * @param {object} config - OpenAPI configuration
   * @param {string} [uiPath='/docs'] - Path to serve Swagger UI
   * @returns {Codehooks} this for chaining
   */
  openapi = (config = {}, uiPath = '/docs') => {
    const specPath = config.specPath || '/openapi.json';

    // Store OpenAPI config
    this.set('_openapi_config', config);

    // Route: Serve OpenAPI JSON spec
    this.get(specPath, async (req, res) => {
      const spec = await generateOpenApiSpec(this, config);
      res.set('Content-Type', 'application/json');
      res.json(spec);
    });

    // Route: Serve Swagger UI HTML
    this.get(uiPath, async (req, res) => {
      // Build spec URL relative to current path
      const specUrl = specPath;

      const html = generateSwaggerHtml(specUrl, {
        title: config.info?.title || 'API Documentation',
        ...config.swaggerUi
      });

      res.set('Content-Type', 'text/html');
      res.send(html);
    });

    // Bypass auth for doc routes
    this.auth(specPath, (req, res, next) => next());
    this.auth(uiPath, (req, res, next) => next());

    return this;
  }

  realtime = (path, ...hook) => {
    this.realtime[path] = hook;
    if (hook) {
      this.useRoute(path, ...hook)
    }
    // SSE response
    this.get(`${path}/:clientID`, async (req, res) => {
      res.set('x-realtime-connect', path);
      res.status(200).json({...data});      
    })
  }

  internalFetch = (url, options = {}) => {
    return internalFetch(url, options);
  }
  
  createWorkflow = (name, description, steps, options={}) => {
    const wf = new Workflow(name, description, steps, options, Codehooks.getInstance());
    wf.register(this);
    return wf;
  }

  init = (hook) => {
    const manifest = {
      workerhooks: this.workers,
      routehooks: this.routes,
      queuehooks: this.queues,
      jobhooks: this.jobs,
      authhooks: this.auths,
      globalhooks: this.appmiddleware,
      settings: this.settings,
      realtime: this.realtime,
      app: this
    };
    if (hook) {
      manifest.startup = hook;
    }
    return manifest;
  };
  // alias
  start = this.init;

  setDatastore = (ds) => {
    this.datastore = ds;
    this.listeners.forEach(observer => observer(this))
  }

  getDatastore = () => {
    return {
      open: () => {
        //console.debug("Open datastore");
        return new Promise((resolve) => {
          resolve(this.datastore)
        })
      }
    }
  }

  addListener = (observer) => {
    this.listeners.push(observer);
  }

  static getInstance() {
    if (!this.instance) {
      this.instance = new Codehooks();
    }
    return this.instance;
  }
  
}


const _coho = Codehooks.getInstance();

let _ds = null;


try {
  DB;
  _ds = DB;
} catch (e) {
  if (e instanceof ReferenceError) {
    // Running standalone
    _ds = _coho.getDatastore();
  }
}

export const Datastore = _ds;
export const datastore = _ds;
export const aggregation = agg;
export const crudlify = crud;
export const coho = _coho;
export const app = _coho;
export {
  Workflow
};
// OpenAPI documentation helpers
export { openapi, response, body, param, query, header } from './openapi/index.mjs';
export const realtime = {
  createChannel: (path, ...hook) => {
    Codehooks.getInstance().realtime(path, ...hook);
  },
  publishEvent: async (channel, data, query={}) => {
    const db = await datastore.open();
    return db.insertOne(`_event_publish_${channel.replace('/', '')}`, {channel, data, query, timestamp: new Date().toISOString()})
  },
  createListener: async (channel, data) => {
    const db = await datastore.open();
    return db.insertOne(`_event_listeners_${channel.replace('/', '')}`, {subscribe: data, timestamp: new Date().toISOString()})
  },
  getListeners: async (channel) => {
    const db = await datastore.open();
    return db.find(`_event_listeners_${channel.replace('/', '')}`, {}).toArray();
  },
  getListener: async (channel, listenerID) => {
    const db = await datastore.open();
    return db.findOne(`_event_listeners_${channel.replace('/', '')}`, {_id: listenerID})
  },  
  removeListener: async (channel, listenerID) => {
    const db = await datastore.open();
    return db.removeOne(`_event_listeners_${channel.replace('/', '')}`, {_id: listenerID})
  }
}

export const schedule = {
  runAt: async (date, payload, worker) => {
    const conn = await datastore.open()
    return await conn.runAt(date, payload, worker)
  },
  run: async (payload, worker) => {
    const conn = await datastore.open()
    return await conn.runAt(new Date(), payload, worker)
  },
  cron: async (cronExpression, ...hook) => {
    return Codehooks.getInstance().job(cronExpression, ...hook);
  }
}

const _filestore = {
  getReadStream: (path, options={}) => {
    return _coho_fs.getFileStream(path, options);
  },
  readFile: (path, options={}) => {
    return _coho_fs.readFile(path, options);
  },
  saveFile: (path, buffer, options={}) => {
    return _coho_fs.saveFile(path, buffer, options);
  },
  deleteFile: (path, options={}) => {
    return _coho_fs.deleteFile(path, options);
  },
  list: (path, options={}) => {
    return _coho_fs.list(path, options);
  },
  readFileAsBuffer: (path) => {
    return new Promise(async (resolve, reject) => {
      const filestream = await filestore.getReadStream(path);
      const chunks = [];
      // create a Buffer
      filestream
        .on('data', (buf) => {
          chunks.push(buf);
        })
        .on('end', () => {
          resolve(Buffer.concat(chunks))
        })
    })
  }
}

export const filestore = _filestore;
export default _coho;
