// codehooks-js/webserver.mjs
import mime from 'mime';
import fetch from 'node-fetch';
import { URL } from 'url';

// old and new style support
function promisify(cb, fn) {
    return new Promise(function (resolve, reject) {
        cb = cb || function (err, html) {
            if (err) {
                return reject(err);
            }
            resolve(html);
        };
        fn(cb);
    });
}


const defaultOptions = { route: '/public', default: "index.html", directory: "/" };

export function serveStatic(options, app, filestore, hook, fsoptions) {
    if (!options) {
        options = defaultOptions;
    }
    let baseRoute = '';
    if (options.route && options.route !== undefined && options.route !== '/') {
        baseRoute = options.route;
    }
    // match path and file
    const regex = new RegExp(`^${baseRoute}(\/[\\w\/_-]*|[\\w\/_-]+\/)?([\\w.-]+)?(\\?[\\w=&%-]+)?$`);
    // create public access for static files
    app.auth(regex, (req, res, next) => {
        let filePath = decodeURIComponent(req.apiPath);
        
        if (filePath == options.route) {
            filePath = '/' + (options.default !== undefined ? options.default : 'index.html');
        }
        const type = mime.getType(filePath)
        if (!type) {
            if (options.notFound) {
                // continue to next middleware
                next()
            } else {
                next(`Static server: ${filePath} is not a valid file name`)
            }
        } else {            
            // ok to grant public access to file
            next()
        }        
    })

    const readFunc = async (req, res) => {
        let filePath = decodeURIComponent(req.apiPath);
        try {
            // serve root '' or '/'
            if (filePath == options.route) {
                filePath = '/' + (options.default !== undefined ? options.default : 'index.html');
            }
            // remove route from file path
            if (options.route && options.route !== undefined && options.route !== '/') {
                filePath = filePath.replace(options.route, '');
            }
            // add directory filesystem path
            if (options.directory && options.directory !== undefined && options.directory !== '/') {
                filePath = options.directory + filePath;
            }
            // serve index.html for empty sub routes '/a/b/'
            if(filePath.endsWith('/')) {
                filePath += (options.default !== undefined ? options.default : 'index.html')
            }

            // get mime type
            const type = mime.getType(filePath)
            
            const filestream = await filestore.getReadStream(filePath, fsoptions);
            console.debug('Serve file', filePath, type)
            res.set('Content-Type', type || 'application/octet-stream');
            if (options.headers) {
                res.headers(options.headers);
                /* removed this default cache setting
                const ONE_DAY =  86400000;
                res.set('Cache-Control', `public, max-age=${ONE_DAY*15}, s-maxage=${ONE_DAY*15}`);
                res.setHeader("Expires", new Date(Date.now() + ONE_DAY*15).toUTCString());
                res.removeHeader('Pragma');
                */
            }            
            // stream content back to client    
            filestream
                .on('data', (buf) => {
                    res.write(buf, 'buffer')
                })
                .on('end', () => {
                    res.end()
                })

        } catch (error) {
            console.debug("Error serving file", filePath, error.message)
            if (options.notFound) {
                // read default file
                // add directory filesystem path
                filePath = options.notFound;
                if (options.directory && options.directory !== undefined && options.directory !== '/') {
                    filePath = options.directory + filePath;
                }
                console.debug('Read notFound file', filePath)
                const type = mime.getType(filePath)
                const filestream = await filestore.getReadStream(filePath, fsoptions);
                res.set('Content-Type', type || 'application/octet-stream');
                res.status(404);
                filestream
                .on('data', (buf) => {
                    res.write(buf, 'buffer')
                })
                .on('end', () => {
                    res.end()
                })
            } else {
                res.status(404).end(`No file here: ${filePath} (${error.message})`)
            }
        }
    }
    const farr = [readFunc]
    if (typeof hook === 'function'){
        farr.unshift(hook)
    } 
    app.get(regex, ...farr)
}

/* template engine support */

let tmplCache = {};


export async function render(viewFile, data, settings, cb) {
    const engineSetting = settings['view engine'];
    const engineName = Object.keys(engineSetting)[0];
    //console.log('View engine', engineName)
    if (!engineSetting) return cb("Missing view engine");

    switch (engineName) {
        case 'ejs':
            ejs(viewFile, data, settings, cb)
            break;
        case 'handlebars':
        case 'hbs':
            handlebars(viewFile, data, settings, cb)
            break;
        default:
            cb("No view engine defined for "+engineName)
            break;
    }
}

async function ejs(viewFile, data, settings, cb) {
    try {
        //console.log('EJS engine render', viewFile, settings, cb)
        let engine = settings['view engine'].ejs;
        let layout = settings['layout'];
        let views = settings['views'];
        let tmpl = null;
        if (tmplCache[viewFile]) {                        
            tmpl = tmplCache[viewFile];
        } else {
            const txtfile = await _coho_fs.readFile(`${settings.views}/${viewFile}.ejs`, {source: true});
            //console.log('Txtfile', txtfile)
            tmplCache[viewFile] = engine.compile(txtfile);
            tmpl = tmplCache[viewFile];
        }                
        cb(null, tmpl(data));        
    } catch (err) {
        cb(err);
    }
}

let layoutFile = null;
async function handlebars(viewFile, data, settings, cb) {
    try {
        const db = await Datastore.open();
        let engine = settings['view engine'].hbs || settings['view engine'].handlebars;  
        let layout = settings['layout']; 
        let views = settings['views'];     
        let tmpl = null;
        //console.log('Hbs view init', layout, views)
        if (layout) {
            const layoutPath = `${layout}`;
            
            if (!layoutFile) {
                
                const cachedLayout = await db.get(`hbs:layout:${layout}`);
                if (cachedLayout) {
                    console.debug('Layout load from cache', layoutPath)
                    layoutFile = cachedLayout;
                } else {
                    console.debug('Layout load from file', layoutPath)
                    layoutFile = await _coho_fs.readFile(layoutPath, {source: true});
                    db.set(`hbs:layout:${layout}`, layoutFile);
                }
            }
            //console.log('Layout file', layoutFile)
            engine.registerPartial('layout', layoutFile);
        }
        let txtfile = await db.get(`hbs:${viewFile}`);
        
        if (tmplCache[viewFile] !== undefined) {            
            console.debug('Render view from cache', tmplCache[viewFile])
            tmpl = tmplCache[viewFile];
            return cb(null, tmpl(data));        
        }
        if (txtfile) {
            console.debug('Render view from db')
            tmplCache[viewFile] = engine.compile(txtfile);
            tmpl = tmplCache[viewFile];
            return cb(null, tmpl(data));        
        }
         
                    
        txtfile = await _coho_fs.readFile(`${settings.views}/${viewFile}.hbs`, {source: true});
        console.debug('Creating hbs template', viewFile)
        tmplCache[viewFile] = engine.compile(txtfile);
        tmpl = tmplCache[viewFile];
        db.set(`hbs:${viewFile}`, txtfile);             
        cb(null, tmpl(data));        
    } catch (err) {
        console.error('Handle bars engine error:', err)
        cb(err);
    }
}

// Generic fetch wrapper that routes through apigateway-service
export const internalFetch = (url, options = {}) => {
    // Parse the original URL to extract domain and path
    
    const urlObj = new URL(url);
    const originalHost = urlObj.host;
    const pathAndQuery = urlObj.pathname + urlObj.search;
    
    // Replace domain with apigateway-service
    const proxyUrl = `http://apigateway-service:8888${pathAndQuery}`;
    
    // Merge headers with original host
    const headers = {
      ...options.headers,
      'host': originalHost
    };
    
    // Create new options with updated headers
    const newOptions = {
      ...options,
      headers
    };
    
    return fetch(proxyUrl, newOptions);
  };
