// codehooks-js/openapi/index.mjs
/**
 * OpenAPI documentation module for Codehooks
 *
 * @example
 * import { app, openapi } from 'codehooks-js';
 *
 * // Enable OpenAPI docs
 * app.openapi({ info: { title: 'My API', version: '1.0.0' } }, '/docs');
 *
 * // Document a custom route
 * app.get('/health',
 *   openapi({
 *     summary: 'Health check',
 *     tags: ['System'],
 *     responses: { 200: { description: 'OK' } }
 *   }),
 *   (req, res) => res.json({ status: 'ok' })
 * );
 */

export { generateOpenApiSpec } from './generator.mjs';
export { generateSwaggerHtml } from './swagger-ui.mjs';
export { convertToJsonSchema, detectSchemaType } from './schema-converter.mjs';
export { generateCrudlifyPaths } from './crudlify-docs.mjs';

/**
 * Create an OpenAPI documentation wrapper middleware
 *
 * This middleware attaches OpenAPI metadata to the route, which is then
 * extracted during route registration and used for spec generation.
 *
 * @param {object} spec - OpenAPI operation specification
 * @param {string} [spec.summary] - Short summary of the operation
 * @param {string} [spec.description] - Detailed description
 * @param {string[]} [spec.tags] - Tags for grouping in docs
 * @param {string} [spec.operationId] - Unique operation identifier
 * @param {Array} [spec.parameters] - Path, query, header, cookie parameters
 * @param {object} [spec.requestBody] - Request body specification
 * @param {object} [spec.responses] - Response specifications by status code
 * @param {Array} [spec.security] - Security requirements
 * @param {boolean} [spec.deprecated] - Whether operation is deprecated
 * @returns {Function} Middleware function with attached metadata
 *
 * @example
 * app.post('/users',
 *   openapi({
 *     summary: 'Create a new user',
 *     tags: ['Users'],
 *     requestBody: {
 *       required: true,
 *       content: {
 *         'application/json': {
 *           schema: {
 *             type: 'object',
 *             required: ['name', 'email'],
 *             properties: {
 *               name: { type: 'string' },
 *               email: { type: 'string', format: 'email' }
 *             }
 *           }
 *         }
 *       }
 *     },
 *     responses: {
 *       201: { description: 'User created' },
 *       400: { description: 'Validation error' }
 *     }
 *   }),
 *   async (req, res) => {
 *     // Handle user creation
 *   }
 * );
 */
export function openapi(spec) {
  // Create a pass-through middleware
  const wrapper = function openapiMiddleware(req, res, next) {
    // Simply continue to the next handler
    if (typeof next === 'function') {
      next();
    }
  };

  // Attach the OpenAPI spec to the function for later extraction
  wrapper._openApiSpec = spec;

  // Mark as openapi middleware for easy detection
  wrapper._isOpenApiMiddleware = true;

  return wrapper;
}

/**
 * Helper to create response specification
 * @param {string} description - Response description
 * @param {object} [schema] - Response schema (JSON Schema or Zod/Yup)
 * @param {string} [contentType='application/json'] - Content type
 * @returns {object} OpenAPI response object
 *
 * @example
 * openapi({
 *   responses: {
 *     200: response('Success', UserSchema),
 *     404: response('Not found')
 *   }
 * })
 */
export function response(description, schema, contentType = 'application/json') {
  if (!schema) {
    return { description };
  }

  return {
    description,
    content: {
      [contentType]: { schema }
    }
  };
}

/**
 * Helper to create request body specification
 * @param {object} schema - Request body schema (JSON Schema or Zod/Yup)
 * @param {object} [options] - Additional options
 * @param {boolean} [options.required=true] - Whether body is required
 * @param {string} [options.description] - Body description
 * @param {string} [options.contentType='application/json'] - Content type
 * @returns {object} OpenAPI request body object
 *
 * @example
 * openapi({
 *   requestBody: body(CreateUserSchema, { description: 'User data' })
 * })
 */
export function body(schema, options = {}) {
  const {
    required = true,
    description,
    contentType = 'application/json'
  } = options;

  return {
    required,
    ...(description && { description }),
    content: {
      [contentType]: { schema }
    }
  };
}

/**
 * Helper to create path parameter specification
 * @param {string} name - Parameter name
 * @param {object} [options] - Parameter options
 * @param {string} [options.description] - Parameter description
 * @param {object} [options.schema] - Parameter schema
 * @param {*} [options.example] - Example value
 * @returns {object} OpenAPI parameter object
 *
 * @example
 * openapi({
 *   parameters: [
 *     param('id', { description: 'User ID', schema: { type: 'string', format: 'uuid' } })
 *   ]
 * })
 */
export function param(name, options = {}) {
  return {
    name,
    in: 'path',
    required: true,
    schema: options.schema || { type: 'string' },
    ...(options.description && { description: options.description }),
    ...(options.example !== undefined && { example: options.example })
  };
}

/**
 * Helper to create query parameter specification
 * @param {string} name - Parameter name
 * @param {object} [options] - Parameter options
 * @param {string} [options.description] - Parameter description
 * @param {boolean} [options.required=false] - Whether required
 * @param {object} [options.schema] - Parameter schema
 * @param {*} [options.example] - Example value
 * @returns {object} OpenAPI parameter object
 *
 * @example
 * openapi({
 *   parameters: [
 *     query('limit', { description: 'Max items', schema: { type: 'integer', default: 10 } }),
 *     query('search', { description: 'Search term' })
 *   ]
 * })
 */
export function query(name, options = {}) {
  return {
    name,
    in: 'query',
    required: options.required || false,
    schema: options.schema || { type: 'string' },
    ...(options.description && { description: options.description }),
    ...(options.example !== undefined && { example: options.example })
  };
}

/**
 * Helper to create header parameter specification
 * @param {string} name - Header name
 * @param {object} [options] - Parameter options
 * @returns {object} OpenAPI parameter object
 */
export function header(name, options = {}) {
  return {
    name,
    in: 'header',
    required: options.required || false,
    schema: options.schema || { type: 'string' },
    ...(options.description && { description: options.description }),
    ...(options.example !== undefined && { example: options.example })
  };
}
