// codehooks-js/workflow/engine.mjs
/*
Implements a workflow engine that uses Codehooks.io as the state storage and queues for persistent workers.
The engine manages step transitions, state persistence, and event handling for workflow-based applications.
*/

import { EventEmitter } from 'events';


/**
 * Workflow class that manages step-based workflows
 * @extends EventEmitter
 */
class Workflow extends EventEmitter {

    // Protected instance variables
    #collectionName = 'workflowdata'; // collection name for storing workflow data
    #queuePrefix = 'workflowqueue'; // queue prefix for storing workflow data
    #timeout = 30000; // timeout for a workflow instance
    #maxStepCount = 3; // maximum number of steps to execute
    #steps = {}; // workflowsteps configuration
    #definitions = null; // workflow definition map
    #name = null; // workflow name
    #description = null; // workflow description
    #defaultStepOptions = { // default step options
        timeout: 30000, // timeout for a step
        maxRetries: 3, // maximum number of retries for a step
        workers: 1 // number of workers for a step
    };

    /**
     * Create a new workflow instance
     * @param {string} name - Unique identifier for the workflow
     * @param {string} description - Human-readable description
     * @param {Object} definition - Object containing step definitions
     * @param {Object} options - Optional configuration options
     */
    constructor(name, description, definition, options = {}, app) {
        super();
        this.#definitions = new Map();
        this.#name = name;
        this.#description = description;

        // Apply any configuration options
        if (options) {
            this.configure(options);
        }

        // Store the initial definition if provided
        if (definition) {
            this.#definitions.set(name, definition);
            console.debug('Initial workflow definition stored:', {
                name,
                stepNames: Object.keys(definition)
            });
        }
    }

    /**
     * Set the collection name for storing steps data
     * @param {string} name - Collection name
     * @throws {Error} If name is not a non-empty string
     */
    setCollectionName(name) {
        if (typeof name !== 'string' || !name.trim()) {
            throw new Error('Collection name must be a non-empty string');
        }
        this.#collectionName = name.trim();
    }

    /**
     * Get the collection name
     * @returns {string} The collection name
     */
    getCollectionName() {
        return this.#collectionName;
    }

    /**
     * Set the queue prefix for steps jobs
     * @param {string} prefix - Queue prefix
     * @throws {Error} If prefix is not a non-empty string
     */
    setQueuePrefix(prefix) {
        if (typeof prefix !== 'string' || !prefix.trim()) {
            throw new Error('Queue prefix must be a non-empty string');
        }
        this.#queuePrefix = prefix.trim();
    }

    /**
     * Get the queue prefix
     * @returns {string} The queue prefix
     */
    getQueuePrefix() {
        return this.#queuePrefix;
    }

    /**
     * Set the timeout for steps jobs
     * @param {number} timeout - Timeout in milliseconds
     * @throws {Error} If timeout is not a positive number
     */
    setTimeout(timeout) {
        if (typeof timeout !== 'number' || timeout <= 0) {
            throw new Error('Timeout must be a positive number');
        }
        this.#timeout = timeout;
    }

    /**
     * Get the timeout
     * @returns {number} The timeout in milliseconds
     */
    getTimeout() {
        return this.#timeout;
    }

    /**
     * Set the maximum step count for a steps workflow
     * @param {number} maxStepCount - Maximum step count
     * @throws {Error} If maxStepCount is not a positive number
     */
    setMaxStepCount(maxStepCount) {
        if (typeof maxStepCount !== 'number' || maxStepCount <= 0) {
            throw new Error('Maximum step count must be a positive number');
        }
        this.#maxStepCount = maxStepCount;
    }

    /**
     * Get the maximum step count
     * @returns {number} The maximum step count
     */
    getMaxStepCount() {
        return this.#maxStepCount;
    }

    /**
     * Set the steps configuration
     * @param {Object} steps - Steps configuration
     * @throws {Error} If steps is not an object or is null
     */
    setStepsConfig(steps) {
        if (typeof steps !== 'object' || steps === null) {
            throw new Error('Steps must be an object');
        }
        this.#steps = steps;
    }

    /**
     * Get the steps configuration
     * @returns {Object} The steps configuration
     */
    getStepsConfig() {
        return this.#steps;
    }

    /**
     * Get the number of workers for a specific step
     * @param {string} stepName - Name of the step
     * @returns {number} The number of workers for the step
     */
    getWorkersForStep(stepName) {
        const stepConfig = this.#steps[stepName];
        return stepConfig?.workers ?? this.#defaultStepOptions.workers ?? 1;
    }

    /**
     * Get the timeout for a specific step
     * @param {string} stepName - Name of the step
     * @returns {number} The timeout in milliseconds for the step
     */
    getTimeoutForStep(stepName) {
        const stepConfig = this.#steps[stepName];
        return stepConfig?.timeout ?? this.#defaultStepOptions.timeout ?? this.#timeout;
    }

    /**
     * Set the default number of workers for steps
     * @param {number} workers - Number of workers
     * @throws {Error} If workers is not a positive number
     */
    setDefaultWorkers(workers) {
        if (typeof workers !== 'number' || workers <= 0) {
            throw new Error('Workers must be a positive number');
        }
        this.#defaultStepOptions.workers = workers;
    }

    /**
     * Get the default number of workers
     * @returns {number} The default number of workers
     */
    getDefaultWorkers() {
        return this.#defaultStepOptions.workers;
    }

    /**
     * Set the default timeout for steps
     * @param {number} timeout - Timeout in milliseconds
     * @throws {Error} If timeout is not a positive number
     */
    setDefaultTimeout(timeout) {
        if (typeof timeout !== 'number' || timeout <= 0) {
            throw new Error('Timeout must be a positive number');
        }
        this.#defaultStepOptions.timeout = timeout;
    }

    /**
     * Get the default timeout for steps
     * @returns {number} The default timeout in milliseconds
     */
    getDefaultTimeout() {
        return this.#defaultStepOptions.timeout;
    }

    /**
     * Configure the steps engine
     * @param {Object} config - Configuration object
     * @param {string} config.collectionName - Collection name
     * @param {string} config.queuePrefix - Queue prefix
     * @param {number} config.timeout - Timeout in milliseconds
     * @param {number} config.maxStepCount - Maximum step count
     * @param {number} config.workers - Default number of workers for steps
     * @param {Object} config.steps - Steps configuration
     */
    configure(config) {
        if (config.collectionName) {
            this.setCollectionName(config.collectionName);
        }
        if (config.queuePrefix) {
            this.setQueuePrefix(config.queuePrefix);
        }
        if (config.timeout) {
            this.setTimeout(config.timeout);
            this.setDefaultTimeout(config.timeout);
        }
        if (config.maxStepCount) {
            this.setMaxStepCount(config.maxStepCount);
        }
        if (config.workers) {
            this.setDefaultWorkers(config.workers);
        }
        if (config.steps) {
            this.setStepsConfig(config.steps);
        }
    }

    /**
     * Get the step definition for a specific step
     * @param {string} stepsName - Name of the steps workflow
     * @param {string} stepName - Name of the step
     * @returns {Function} The step function
     * @throws {Error} If steps definition or step function not found
     */
    getDefinition(stepsName, stepName) {
        const stepsDef = this.#definitions.get(stepsName);
        if (!stepsDef) {
            throw new Error(`No Steps definition found for: ${stepsName}`);
        }
        // Get the specific step function from that workflow
        const stepFunc = stepsDef[stepName];
        if (!stepFunc) {
            throw new Error(`No step function found for: ${stepName} in steps ${stepsName}`);
        }
        return stepFunc;
    }

    /**
     * Validate a step definition
     * @param {Function} step - The step function to validate
     * @throws {Error} If step definition is invalid
     */
    validateStepDefinition(step) {
        if (typeof step !== 'function') {
            throw new Error('Step definition must be a function');
        }
        if (step.length !== 2) {
            throw new Error('Step definition must accept exactly 2 parameters');
        }
    }

    /**
     * Handle the next step in the workflow
     * @param {string} stepsName - Name of the steps workflow
     * @param {string|null} nextStep - Name of the next step or null if finished
     * @param {Object} newState - The new state to set
     * @param {string} instanceId - ID of the steps instance
     * @param {Object} options - Step options (waitForInput, timeout, retry, abort)
     * @returns {Promise<string|string[]>} Queue ID(s) for the next step(s)
     * @throws {Error} If step execution fails
     */
    async handleNextStep(stepsName, nextStep, newState, instanceId, options) {

        // open the connection to the database
        const connection = await DB.open();

        // Handle single next step
        this.emit('stepStarted', { workflowName: stepsName, step: nextStep, state: newState, instanceId });

        // remove the _id from the newState
        delete newState._id;
        // increment the step count
        if (!newState.stepCount) {
            newState.stepCount = {};
        }
        if (!newState.stepCount[nextStep]) {
            newState.stepCount[nextStep] = { visits: 0, startTime: new Date().toISOString(), totalTime: 0 };
        }
        newState.stepCount[nextStep].visits += 1;
        // Update the existing steps state in the database
        newState = await connection.updateOne(this.#collectionName,
            { _id: instanceId },
            { $set: { ...newState, nextStep: nextStep, updatedAt: new Date().toISOString(), stepCount: newState.stepCount } });

        this.emit('stateUpdated', { workflowName: stepsName, state: newState, instanceId });

        try {
            // Get the next step function
            const func = this.getDefinition(stepsName, nextStep);

            // Wrap the callback in a Promise to ensure proper async handling
            return await new Promise(async (resolve, reject) => {
                // check if the step count is greater than the max step count
                if (newState.stepCount[nextStep].visits > this.getMaxStepCount()) {
                    reject(new Error(`Step ${nextStep} has been executed ${newState.stepCount[nextStep].visits} times, which is greater than the maximum step count of ${this.getMaxStepCount()}`));
                    return;
                }
                try {
                    const waiterFunction = async (state) => {
                        if (state) {
                            console.debug('waiter', state);
                            await connection.updateOne(this.#collectionName,
                                { _id: instanceId },
                                { $set: { ...state, updatedAt: new Date().toISOString() } });
                            resolve();
                            return;
                        } else {
                            resolve(); // no state update
                        }
                    }
                    await func({ ...newState, instanceId: newState._id }, async (nextStep, userState, options) => {
                        try {
                            // Protect system-level properties
                            const protectedState = {
                                _id: newState._id,
                                nextStep: newState.nextStep,
                                createdAt: newState.createdAt,
                                updatedAt: newState.updatedAt,
                                instanceId: newState.instanceId,
                                workflowName: newState.workflowName,
                                stepCount: newState.stepCount,
                                parallelSteps: newState.parallelSteps,
                                previousStep: newState.nextStep
                            };

                            // Merge states with userState taking precedence, but protecting system fields
                            const mergedState = {
                                ...userState,
                                ...protectedState
                            };

                            // set step finish time
                            mergedState.stepCount[mergedState.nextStep].finishTime = new Date().toISOString();
                            mergedState.stepCount[mergedState.nextStep].totalTime = (new Date(mergedState.stepCount[mergedState.nextStep].finishTime) - new Date(mergedState.stepCount[mergedState.nextStep].startTime));

                            // update the parallel steps metadata
                            if (mergedState.parallelSteps && mergedState.parallelSteps[mergedState.nextStep]) {
                                // get a fresh copy of the parallel steps
                                const fresh = await connection.findOne(this.#collectionName, { _id: instanceId });
                                fresh.parallelSteps[mergedState.nextStep].done = true;
                                fresh.parallelSteps[mergedState.nextStep].nextStep = nextStep;
                                //fresh.parallelSteps[mergedState.nextStep].previousStep = mergedState.previousStep || null;                           
                                delete fresh._id;
                                const updated = await connection.updateOne(this.#collectionName,
                                    { _id: instanceId },
                                    { $set: { ...fresh, parallelSteps: fresh.parallelSteps } });
                                //console.debug('updated', updated.parallelSteps);
                                mergedState.parallelSteps = fresh.parallelSteps;
                                // Check if all parallel steps are done
                                const allStepsDone = Object.values(fresh.parallelSteps).every(step => step.done);
                                if (!allStepsDone) {
                                    console.debug('Waiting for other parallel steps to complete');
                                    resolve();
                                    return;
                                } else {
                                    // validate that all parallel steps have the same next step
                                    const nextSteps = Object.values(fresh.parallelSteps).map(step => step.nextStep);
                                    const uniqueNextSteps = [...new Set(nextSteps)];
                                    if (uniqueNextSteps.length > 1) {
                                        throw new Error('Parallel steps must join to the same next step');
                                    }
                                    // reset the parallel steps
                                    mergedState.parallelSteps = {};
                                    console.debug('All parallel steps are done');
                                }
                            }

                            // If there is no next step, the workflow is completed
                            if (nextStep === null) {
                                delete mergedState._id;
                                const completionTime = (new Date(mergedState.updatedAt) - new Date(mergedState.createdAt));

                                const finalresult = await connection.updateOne(this.#collectionName,
                                    { _id: instanceId },
                                    { $set: { ...mergedState, nextStep: null, updatedAt: new Date().toISOString(), totalTime: completionTime } });
                                console.log(`Workflow ${stepsName} ${instanceId} is completed in time: ${completionTime / 1000}s 🎉`);
                                this.emit('completed', { ...finalresult });
                                resolve();
                                return;
                            }

                            // Additional safety check to prevent enqueueing null or invalid steps
                            if (nextStep === null || nextStep === undefined || nextStep === 'null') {
                                throw new Error(`Invalid nextStep value: ${nextStep}. This should have been caught earlier.`);
                            }

                            // Enqueue the next step, single or parallel                            
                            if (Array.isArray(nextStep)) {
                                const now = new Date().toISOString();
                                const metadata = nextStep.reduce((acc, step) => {
                                    acc[step] = { done: false, startTime: now, previousStep: mergedState.previousStep };
                                    return acc;
                                }, {});
                                const metadataDoc = await connection.updateOne(this.#collectionName,
                                    { _id: instanceId },
                                    { $set: { parallelSteps: metadata } });
                                //console.log('metadataDoc', metadataDoc);
                                // enqueue all steps in parallel
                                for (const step of nextStep) {
                                    console.debug('enqueue step', step, instanceId);
                                    await connection.enqueue(`${this.#queuePrefix}_${stepsName}_${step}`, {
                                        stepsName: stepsName,
                                        goto: step,
                                        state: mergedState,
                                        options: options,
                                        instanceId: instanceId
                                    }, { workers: this.getWorkersForStep(step), timeout: this.getTimeoutForStep(step) });
                                }
                            } else {
                                console.debug('enqueue step', nextStep, instanceId);
                                await connection.enqueue(`${this.#queuePrefix}_${stepsName}_${nextStep}`, {
                                    stepsName: stepsName,
                                    goto: nextStep,
                                    state: mergedState,
                                    options: options,
                                    instanceId: instanceId
                                }, { workers: this.getWorkersForStep(nextStep), timeout: this.getTimeoutForStep(nextStep) });
                            }
                            this.emit('stepEnqueued', { workflowName: stepsName, step: nextStep, state: newState, instanceId });
                            resolve();
                        } catch (error) {
                            console.error('error', error.message);
                            reject(error);
                        }
                    }, waiterFunction);
                    //resolve();
                } catch (error) {
                    console.error('Error executing step function:', error);
                    reject(error);
                }
            });
        } catch (error) {
            console.error('error in handleNextStep outer', error.message);
            throw error;
        }
    }

    /**
     * Register the workflow with a Codehooks application
     * @param {Codehooks} app - Codehooks application instance
     * @returns {Promise<string>} The registered workflow name
     */
    async register(app) {
        if (!this.#name || !this.#description) {
            throw new Error('Workflow name and description must be set before registration');
        }

        const definition = this.#definitions.get(this.#name);
        if (!definition) {
            throw new Error(`No workflow definition found for: ${this.#name}`);
        }

        return this.registerWithApp(app, this.#name, this.#description, definition);
    }

    /**
     * Register a new workflow with a Codehooks application
     * @param {Codehooks} app - Codehooks application instance
     * @param {string} name - Unique identifier for the workflow
     * @param {string} description - Human-readable description
     * @param {Object} definition - Object containing step definitions
     * @returns {Promise<string>} The registered workflow name
     * @private
     */
    async registerWithApp(app, name, description, definition) {
        this.emit('workflowCreated', { name, description });

        // Log current configuration state at registration time
        console.debug('Workflow registration config:', {
            defaultWorkers: this.#defaultStepOptions.workers,
            defaultTimeout: this.#defaultStepOptions.timeout,
            stepsConfig: this.#steps
        });

        // Validate each step in the definition
        for (const [stepName, step] of Object.entries(definition)) {
            try {
                // Skip null, undefined, or invalid step names
                if (stepName !== undefined && stepName !== null && stepName !== 'null' && stepName.trim() !== '') {
                    const workerCount = this.getWorkersForStep(stepName);
                    const timeoutMs = this.getTimeoutForStep(stepName);
                    console.debug('registering queue for step', `${this.#queuePrefix}_${name}_${stepName}`, { workers: workerCount, timeout: timeoutMs, stepConfig: this.#steps[stepName] });
                    app.worker(`${this.#queuePrefix}_${name}_${stepName}`, async (req, res) => {
                        try {
                            const { stepsName, goto, state, instanceId, options } = req.body.payload;
                            console.debug('dequeue step', stepName, instanceId);
                            const qid = await this.handleNextStep(stepsName, goto, state, instanceId, options);
                            //console.debug('Worker qid', qid);
                            //res.end();
                        } catch (error) {
                            const { stepsName, goto, state, instanceId, options } = req.body.payload;
                            console.debug('DB in workflow 1', DB);
                            const connection = await DB.open();
                            await connection.updateOne(this.#collectionName,
                                { _id: instanceId },
                                { $set: { lastError: error, updatedAt: new Date().toISOString() } });
                            console.error('Error in function: ' + stepName, error);
                            this.emit('error', error);
                        } finally {
                            //console.debug('Worker res.end', stepName);
                            res.end();
                        }
                    }, { workers: workerCount, timeout: timeoutMs });
                }
            } catch (error) {
                console.error(`Invalid step definition '${stepName}': ${error.message}`);
                throw error;
            }
        }

        // Store the definition
        this.#definitions.set(name, definition);

        return name;
    }

    /**
     * Start a new steps instance
     * @param {Object} initialState - Initial state for the steps instance
     * @returns {Promise<{id: string}>} The steps instance ID
     * @throws {Error} If starting steps fails
     */
    async start(initialState) {
        this.emit('workflowStarted', { name: this.#name, initialState });

        return new Promise(async (resolve, reject) => {
            try {
                console.debug('Starting workflow', this.#name);
                const funcs = this.#definitions.get(this.#name);

                if (!funcs) {
                    reject(new Error(`No workflow definition found for: ${this.#name}`));
                    return;
                }

                const firstStepName = Object.keys(funcs)[0];
                const firstStep = funcs[firstStepName];

                if (!firstStep) {
                    reject(new Error('No start step defined in workflow'));
                    return;
                }
                console.debug('DB in workflow', DB);
                const connection = await DB.open();
                // Create a new workflow state in the database
                const newState = await connection.insertOne(this.#collectionName,
                    { ...initialState, nextStep: firstStepName, createdAt: new Date().toISOString(), workflowName: this.#name, stepCount: {} });
                const { _id: ID } = await connection.enqueue(`${this.#queuePrefix}_${this.#name}_${firstStepName}`, {
                    stepsName: this.#name,
                    goto: firstStepName,
                    state: newState,
                    instanceId: newState._id,
                    options: {}
                }, { workers: this.getWorkersForStep(firstStepName), timeout: this.getTimeoutForStep(firstStepName) });
                resolve(newState);
            } catch (error) {
                console.error('Error starting workflow:', error.message);
                reject(error);
            }
        });
    }

    /**
     * Update the state of a workflow instance
     * @param {string} instanceId - ID of the workflow instance
     * @param {Object} state - New state to update with
     * @param {Object} options - Options for the update, { continue: false } to avoid continuing the the step
     * @returns {Promise<Object>} The updated state
     */
    async updateState(instanceId, state, options = { continue: true }) {
        this.emit('stepsStateUpdating', { name: this.#name, instanceId, state });
        const connection = await DB.open();
        return new Promise(async (resolve, reject) => {
            const doc = await connection.updateOne(this.#collectionName,
                { _id: instanceId },
                { $set: { ...state, updatedAt: new Date().toISOString() } });
            if (options.continue) {
                await this.continue(instanceId, false);
            }
            resolve({ ...doc });
        });
    }

    /**
     * Set the complete state of a steps instance
     * @param {string} instanceId - ID of the steps instance
     * @param {Object} state - Complete state to set
     * @returns {Promise<void>}
     */
    async setState(instanceId, { _id, state }) {
        const connection = await DB.open();
        await connection.replaceOne(this.#collectionName, { _id: _id }, { ...state });
    }

    /**
     * Continue a paused steps instance
     * @param {string} instanceId - ID of the steps instance
     * @returns {Promise<{qId: string}>} Queue ID for the continued step
     * @throws {Error} If steps instance not found
     */
    async continue(instanceId, reset = false) {
        const connection = await DB.open();
        const state = await connection.findOne(this.#collectionName, { _id: instanceId });
        if (!state) {
            throw new Error(`No steps found with instanceId: ${instanceId}`);
        }
        
        // Check if the workflow has already completed
        if (state.nextStep === null) {
            throw new Error(`Workflow ${instanceId} has already completed and cannot be continued`);
        }
        
        if (reset) {
            // reset the step count
            // update all step counts to 0  
            for (const step in state.stepCount) {
                state.stepCount[step] = { visits: 0, startTime: new Date().toISOString() };
            }
        } else {
            // update the step count
            state.stepCount[state.nextStep] = { visits: 0, startTime: new Date().toISOString() };
        }

        await connection.updateOne(this.#collectionName, { _id: instanceId }, { $set: { stepCount: state.stepCount } });
        console.debug('continue state', state);
        this.emit('workflowContinued', { name: this.#name, step: state.nextStep, instanceId });

        return new Promise(async (resolve, reject) => {
            const { _id: ID } = await connection.enqueue(`${this.#queuePrefix}_${this.#name}_${state.nextStep}`, {
                stepsName: this.#name,
                goto: state.nextStep,
                state: state,
                options: {},
                instanceId: instanceId
            }, { workers: this.getWorkersForStep(state.nextStep), timeout: this.getTimeoutForStep(state.nextStep) });

            resolve({ instanceId });
        });
    }

    /**
     * Continue all timed out workflows instances
     * @returns {Promise<Array<{qId: string}>>} Array of results containing queue IDs for continued workflows
     */
    async continueAllTimedOut() {
        const db = await DB.open();
        const timedOutWorkflows = await db.collection(this.#collectionName).find({ nextStep: { $ne: null } }).toArray();
        const now = new Date();
        const results = [];
        for (const workflow of timedOutWorkflows) {
            const createdAt = new Date(workflow.createdAt);
            const diffMillis = now.getTime() - createdAt.getTime();
            if (diffMillis > this.#timeout) {
                const diffMinutes = diffMillis / (1000 * 60);
                console.log('Timed out:', workflow._id, workflow.nextStep, `(${diffMinutes.toFixed(1)} minutes old)`);
                const result = await this.continue(workflow._id, true);
                console.log('Continued:', result._id);
                results.push(result);
            }
        }
        return results;
    }

    /**
     * Get the status of a steps instance
     * @param {string} id - ID of the steps instance
     * @returns {Promise<Object>} The steps status
     */
    async getStepsStatus(id) {
        return new Promise(async (resolve, reject) => {
            const connection = await DB.open();
            const state = await connection.findOne(this.#collectionName, { _id: id });
            resolve(state);
        });
    }

    /**
     * List all steps instances matching a filter
     * @param {Object} filter - Filter criteria for steps workflows
     * @returns {Promise<Array>} List of steps instances
     */
    async getInstances(filter) {
        return new Promise(async (resolve, reject) => {
            const connection = await DB.open();
            const states = await connection.find(this.#collectionName, filter).toArray();
            console.debug('listSteps', this.#collectionName, filter, states.length);
            resolve(states);
        });
    }

    /**
     * Cancel a steps instance
     * @param {string} id - ID of the steps instance to cancel
     * @returns {Promise<Object>} The cancellation result
     */
    async cancelSteps(id) {
        this.emit('cancelled', { id });
        return new Promise(async (resolve, reject) => {
            const connection = await DB.open();
            const state = await connection.updateOne(this.#collectionName,
                { _id: id },
                { $set: { status: 'cancelled' } });
            resolve(state);
        });
    }

    /**
     * Get the status of a workflow instance (alias for getStepsStatus)
     * @param {string} id - ID of the workflow instance
     * @returns {Promise<Object>} The workflow status
     */
    async getWorkflowStatus(id) {
        return this.getStepsStatus(id);
    }

    /**
     * Get the state of a workflow instance (alias for getStepsStatus)
     * @param {string} id - ID of the workflow instance
     * @returns {Promise<Object>} The workflow state
     */
    async getState(id) {
        return this.getStepsStatus(id);
    }

    /**
     * Cancel a workflow instance (alias for cancelSteps)
     * @param {string} id - ID of the workflow instance to cancel
     * @returns {Promise<Object>} The cancellation result
     */
    async cancelWorkflow(id) {
        return this.cancelSteps(id);
    }

    /**
     * Check if a specific step in a workflow instance has timed out
     * @param {Object} workflow - The workflow instance
     * @param {string} stepName - The name of the step to check (defaults to previousStep)
     * @returns {Object} Object containing timeout status and details
     */
    isStepTimedOut(workflow) {
        // Use previousStep if no stepName provided
        const stepToCheck = workflow.nextStep;

        if (!stepToCheck || !workflow.stepCount || !workflow.stepCount[stepToCheck]) {
            console.debug('no step', stepToCheck, workflow.stepCount);
            return {
                isTimedOut: false,
                reason: 'Step not found in workflow',
                step: stepToCheck
            };
        }

        const step = workflow.stepCount[stepToCheck];

        // Get the timeout value for this step using the helper method
        const stepConfig = this.#steps[stepToCheck];
        const stepTimeout = this.getTimeoutForStep(stepToCheck);

        // If the step hasn't finished, check if it's been running too long
        console.debug('isStepTimedOut', stepToCheck, stepTimeout);
        if (!step.finishTime) {
            const startTime = new Date(step.startTime);
            const now = new Date();
            const runningTime = now.getTime() - startTime.getTime();
            console.debug('runningTime', runningTime, stepTimeout);
            return {
                isTimedOut: runningTime > stepTimeout,
                runningTime,
                timeout: stepTimeout,
                step: stepToCheck,
                startTime: step.startTime,
                currentTime: now.toISOString(),
                timeoutSource: stepConfig?.timeout ? 'stepConfig' : (this.#defaultStepOptions.timeout !== 30000 ? 'defaultOptions' : 'globalTimeout')
            };
        }

        // If the step has finished, check if it took too long
        const startTime = new Date(step.startTime);
        const finishTime = new Date(step.finishTime);
        const executionTime = finishTime.getTime() - startTime.getTime();
        console.debug('executionTime', executionTime, stepTimeout);
        return {
            isTimedOut: executionTime > stepTimeout,
            executionTime,
            timeout: stepTimeout,
            step: stepToCheck,
            startTime: step.startTime,
            finishTime: step.finishTime,
            timeoutSource: stepConfig?.timeout ? 'stepConfig' : (this.#defaultStepOptions.timeout !== 30000 ? 'defaultOptions' : 'globalTimeout')
        };
    }

    /**
     * Find all workflow instances with timed out steps
     * @param {Object} filter - Optional filter criteria for workflows
     * @returns {Promise<Array<Object>>} Array of workflow instances with timed out steps
     */
    async findTimedOutSteps(filter = {}) {
        const db = await DB.open();
        const workflows = await db.getMany(this.#collectionName, { "nextStep": { $ne: null } }).toArray();
        if (workflows.length > 0) {
            console.debug('TimedOutSteps', workflows.length);
        }
        const timedOutWorkflows = workflows.map(workflow => {
            console.debug('isStepTimedOut', workflow.nextStep);
            const timeoutStatus = this.isStepTimedOut(workflow);
            if (timeoutStatus.isTimedOut) {
                return {
                    workflowId: workflow._id,
                    workflowName: workflow.workflowName,
                    ...timeoutStatus
                };
            }
            return null;
        }).filter(Boolean);

        return timedOutWorkflows;
    }
}

// Export the class directly instead of a singleton instance
export { Workflow };
export default Workflow;
